iceoryx_utils/doc/lockfree_queue/lockfree_queue.md

```diff
@@ -0,0 +1,140 @@
+# Lock Free queue 
+
+We explain some details of the lock free queue in order to provide an intuition for the way the lock free queue works. This could serve as a basis for a formal proof later if desired.
+
+## Index queue analysis
+
+In the following all numbers are natural numbers and hence bounded unsigned integers in the implementation.
+
+### Capacity and Cycle Length
+The index queue has some capacity n and stores indices in the range [0, n[.
+The number n is also called cycle length.
+
+Assume n = 4 in this explanation, so we can store the indices 0, 1, 2, 3.
+
+
+### Index representation
+Each index i corresponds to an equivalence class [i] modulo n, i.e. 
+[i] = {j | j = c*n + i, c >= 0}
+
+This means j = 3, 7, 11, ... all represent the same index 3 (pointing to the last element in the array)
+
+Given j, we have i = j % n and c = j / n (where % is modulo and / integer division).
+
+Another way to represent such a value j uniquely is as a pair (c,i) and
+we call this a cyclic index with cycle c and index i.
+
+This is done for two reasons: detection of empty queues and to eliminate the ABA problem for all practical scenarios.
+It is imperative that such indices can be used in compare and swap operations (CAS), i.e. 
+will not exceed 64 bits on standard architectures supporting 64 bit CAS. 
+Therefore j can be assumed to be an unsigned 64 bit integer.
+
+
+### Queue representation
+The queue has a head H, tail T and an array of n values. All of them are cyclic indices as in 2.
+Initially the queue is empty and head and tail point to index 0 but have both cycle 1. 
+Array elements are 0 and represented as (0,0).
+
+Initially empty queue
+```
+[ (0,0), (0,0), (0,0), (0,0) ]   H=(1,0)     T=(1,0)
+```
+Initially full queue with all indices
+```
+[ (0,0), (0,1), (0,2), (0,3) ]   H=(0,0)     T=(1,0)
+```
+
+
+### Head and Tail monotonicity
+We always insert (push) at Tail and remove (pop) at Head.
+Head and Tail both increase strictly monotonic (making ABA problems unlikely).
+I.e. a push causes tail to increase by one (causing the cycle to increase after n pushes) and similarly each pop causes Head to increase by one.
+
+
+### Push operation
+push(y) causes the cycle at the position Tail points to to be replaced with Tails cycle while the element there is replaced with x
+via a CAS operation.
+Afterwards Tail is increased by 1 (potentially not immediately, but before the next push takes effect).
+```
+[ (c,?), (c,x), (c-1,?), (c-1,?) ]   H=(c,1)     T=(c,2)
+```
+push(y)
+```
+[ (c,?), (c,x), (c,y), (c-1,?) ]   H=(c,1)     T=(c,3)
+```
+
+We only push iff the cycle at the element Tail points to is exactly one behind Tails cycle.

```

Comment:
typo `iff`

Reply:
Not really: https://en.wikipedia.org/wiki/If_and_only_if

Reply:
But may remove it for consistency with everything. This file will be reworked during capacity change feature anyway.